base_layer/core/src/base_node/states/fetching_horizon_state.rs

```diff
@@ -49,60 +81,185 @@ impl HorizonInfo {
+            base_state.combine(leaf_nodes);
+            complete_leaf_count = total_leaf_count as u64;
+            index += leaf_nodes_sync_chunk_size;
+        }
+        Ok(base_state)
+    }
+
+    async fn synchronize_headers<B: BlockchainBackend>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+    ) -> Result<(), String>
+    {
+        let height_indices = (0..self.horizon_block).collect::<Vec<u64>>();
+        for block_nums in height_indices.chunks(shared.config.horizon_sync_config.headers_sync_chunk_size) {
+            let headers = shared
+                .comms
+                .fetch_headers(block_nums.to_vec())
+                .await
+                .map_err(|e| e.to_string())?;
+
+            let mut txn = DbTransaction::new();
+            headers.into_iter().for_each(|header| txn.insert_header(header));
+            shared.db.commit(txn).map_err(|e| e.to_string())?;
+        }
+        Ok(())
     }
 
-    async fn synchronize_kernels(&mut self) -> Result<(), String> {
-        Err("unimplemented".into())
+    async fn synchronize_kernels<B: BlockchainBackend>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+    ) -> Result<(), String>
+    {
+        let mmr_state = shared
+            .db
+            .fetch_mmr_base_leaf_nodes(MmrTree::Kernel, 0, 1)
+            .map_err(|e| e.to_string())?;
+        let kernel_hashes = shared
+            .db
+            .fetch_mmr_base_leaf_nodes(MmrTree::Kernel, 0, mmr_state.total_leaf_count)
+            .map_err(|e| e.to_string())?
+            .leaf_nodes
+            .leaf_hashes;
+
+        for hashes in kernel_hashes.chunks(shared.config.horizon_sync_config.kernels_sync_chunk_size) {
+            let kernels = shared
+                .comms
+                .fetch_kernels(hashes.to_vec())
+                .await
+                .map_err(|e| e.to_string())?;
+
+            let mut txn = DbTransaction::new();
+            kernels.into_iter().for_each(|kernel| txn.insert_kernel(kernel));
+            shared.db.commit(txn).map_err(|e| e.to_string())?;
+        }
+        Ok(())
     }
 
-    async fn synchronize_utxo_set(&mut self) -> Result<(), String> {
-        Err("unimplemented".into())
+    async fn synchronize_utxo_set<B: BlockchainBackend>(
+        &mut self,
+        shared: &mut BaseNodeStateMachine<B>,
+    ) -> Result<(), String>
+    {
+        let mmr_state = shared
+            .db
+            .fetch_mmr_base_leaf_nodes(MmrTree::Utxo, 0, 1)
+            .map_err(|e| e.to_string())?;
+        let MutableMmrLeafNodes {
+            mut leaf_hashes,
+            deleted,
+        } = shared
+            .db
+            .fetch_mmr_base_leaf_nodes(MmrTree::Utxo, 0, mmr_state.total_leaf_count)
+            .map_err(|e| e.to_string())?
+            .leaf_nodes;
+        deleted.to_vec().into_iter().rev().for_each(|index| {
+            let _ = leaf_hashes.remove(index as usize);
+        });
```

Comment:
A thought: If we have 100,000,000 UTXOs, how quickly would this line execute, and how much memory would it consume?

I suspect this would be more efficient, but would need to test
```
let leaf_hashes = leaf_hashes
            .into_iter()
            .enumerate()
            .filter(|(index, h)| !deleted.contains(*index as u32))
            .map(|(_, h)| h)
            .collect();
```

Reply:
I agree, this is an issue. I changed the code so that it doesn't need the entire hash set in memory any more. It will now request a small set of hashes from the local blockchain db that it can fetch from a remote node.  Creating the vector of leaf hashes without the deleted hashes should be much cheaper now as it is performed only on the small hash sets.